File: app/src/main/java/com/stedi/randomimagegenerator/app/presenter/impl/HomePresenterImpl.java
package com.stedi.randomimagegenerator.app.presenter.impl;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.squareup.otto.Subscribe;
import com.stedi.randomimagegenerator.app.BuildConfig;
import com.stedi.randomimagegenerator.app.di.qualifiers.DefaultScheduler;
import com.stedi.randomimagegenerator.app.di.qualifiers.RigScheduler;
import com.stedi.randomimagegenerator.app.di.qualifiers.UiScheduler;
import com.stedi.randomimagegenerator.app.model.data.PendingPreset;
import com.stedi.randomimagegenerator.app.model.data.Preset;
import com.stedi.randomimagegenerator.app.model.repository.PresetRepository;
import com.stedi.randomimagegenerator.app.other.CachedBus;
import com.stedi.randomimagegenerator.app.other.ChainSerializable;
import com.stedi.randomimagegenerator.app.other.logger.Logger;
import com.stedi.randomimagegenerator.app.presenter.interfaces.HomePresenter;

import java.io.Serializable;
import java.util.List;

import rx.Observable;
import rx.Scheduler;

public class HomePresenterImpl extends HomePresenter {
    private final PresetRepository presetRepository;
    private final PendingPreset pendingPreset;
    private final Scheduler subscribeOn;
    private final Scheduler observeOn;
    private final CachedBus bus;
    private final Logger logger;

    private UIImpl ui;
    private boolean fetchInProgress;

    private Confirm lastActionConfirm;
    private int lastActionPresetId;

    private static class FetchPresetsEvent {
        private final List<Preset> presets;
        private final Throwable throwable;

        FetchPresetsEvent(List<Preset> presets, Throwable throwable) {
            this.presets = presets;
            this.throwable = throwable;
        }
    }

    private static class DeletePresetEvent {
        private final Preset preset;
        private final Throwable throwable;

        DeletePresetEvent(Preset preset, Throwable throwable) {
            this.preset = preset;
            this.throwable = throwable;
        }
    }

    public HomePresenterImpl(@NonNull PresetRepository presetRepository, @NonNull PendingPreset pendingPreset,
                             @NonNull @DefaultScheduler Scheduler subscribeOn,
                             @NonNull @RigScheduler Scheduler superSubscribeOn,
                             @NonNull @UiScheduler Scheduler observeOn,
                             @NonNull CachedBus bus, @NonNull Logger logger) {
        super(superSubscribeOn, observeOn, bus, logger);
        this.presetRepository = presetRepository;
        this.pendingPreset = pendingPreset;
        this.subscribeOn = subscribeOn;
        this.observeOn = observeOn;
        this.bus = bus;
        this.logger = logger;
    }

    @Override
    public void fetchPresets() {
        if (!fetchInProgress) {
            logger.log(this, "fetching presets");
            fetchInProgress = true;
            Observable.fromCallable(presetRepository::getAll)
                    .subscribeOn(subscribeOn)
                    .observeOn(observeOn)
                    .subscribe(presets -> bus.post(new FetchPresetsEvent(presets, null)),
                            throwable -> bus.post(new FetchPresetsEvent(null, throwable)));
        }
    }

    @Override
    public void editPreset(@NonNull Preset preset) {
        pendingPreset.prepareCandidateFrom(preset);
        ui.showEditPreset();
    }

    @Override
    public void confirmLastAction() {
        logger.log(this, "confirmLastAction " + lastActionConfirm);
        if (lastActionConfirm == Confirm.DELETE_PRESET) {
            int lastActionPresetIdRef = lastActionPresetId;
            Observable.fromCallable(() -> {
                Preset preset = presetRepository.get(lastActionPresetIdRef);
                presetRepository.remove(lastActionPresetIdRef);
                return preset;
            }).subscribeOn(subscribeOn)
                    .observeOn(observeOn)
                    .subscribe(preset -> bus.post(new DeletePresetEvent(preset, null)),
                            throwable -> bus.post(new DeletePresetEvent(null, throwable)));
        } else if (lastActionConfirm == Confirm.GENERATE_FROM_PRESET) {
            int lastActionPresetIdRef = lastActionPresetId;
            Observable.fromCallable(() -> presetRepository.get(lastActionPresetIdRef))
                    .subscribeOn(subscribeOn)
                    .observeOn(observeOn)
                    .subscribe(super::startGeneration,
                            throwable -> logger.log(this, throwable));
        }
        lastActionConfirm = null;
        lastActionPresetId = 0;
    }

    @Override
    public void cancelLastAction() {
        logger.log(this, "cancelLastAction " + lastActionConfirm);
        lastActionConfirm = null;
        lastActionPresetId = 0;
    }

    @Override
    public void deletePreset(@NonNull Preset preset) {
        if (lastActionConfirm != null) {
            logger.log(this, "ignoring deletePreset, because last action " + lastActionConfirm + " is not confirmed/canceled");
            return;
        }
        logger.log(this, "deletePreset " + preset);
        if (pendingPreset.get() == preset) {
            pendingPreset.clear();
            ui.onPresetDeleted(preset);
            return;
        }
        lastActionConfirm = Confirm.DELETE_PRESET;
        lastActionPresetId = preset.getId();
        ui.showConfirmLastAction(lastActionConfirm);
    }

    @Override
    public void startGeneration(@NonNull Preset preset) {
        if (lastActionConfirm != null) {
            logger.log(this, "ignoring startGeneration, because last action " + lastActionConfirm + " is not confirmed/canceled");
            return;
        }
        logger.log(this, "startGeneration " + preset);
        lastActionConfirm = Confirm.GENERATE_FROM_PRESET;
        lastActionPresetId = preset.getId();
        ui.showConfirmLastAction(lastActionConfirm);
    }

    @Subscribe
    public void onFetchPresetsEvent(FetchPresetsEvent event) {
        logger.log(this, "onFetchPresetsEvent");

        if (!fetchInProgress) {
            logger.log(this, "ignoring event from not retained presenter!");
            return;
        }

        if (ui == null) {
            fetchInProgress = false;
            logger.log(this, "onFetchPresetsEvent when ui == null");
            return;
        }

        if (event.throwable != null) {
            logger.log(this, event.throwable);
            ui.onFailedToFetchPresets();
        } else {
            if (!BuildConfig.BUILD_TYPE.equals("release")) {
                logger.log(this, "onPresetsFetched " + pendingPreset);
                for (Preset preset : event.presets) {
                    logger.log(this, "onPresetsFetched " + preset);
                }
            }
            ui.onPresetsFetched(pendingPreset.get(), event.presets);
        }

        fetchInProgress = false;
    }

    @Subscribe
    public void onDeletePresetEvent(DeletePresetEvent event) {
        logger.log(this, "onDeletePresetEvent");

        if (ui == null) {
            logger.log(this, "onDeletePresetEvent when ui == null");
            return;
        }

        if (event.throwable != null || event.preset == null) {
            logger.log(this, "failed to delete preset", event.throwable);
            ui.onFailedToDeletePreset();
        } else {
            ui.onPresetDeleted(event.preset);
        }
    }

    @Override
    public void onAttach(@NonNull UIImpl ui) {
        super.onAttach(ui);
        this.ui = ui;
        bus.register(this);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        bus.unregister(this);
        ui = null;
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    public void onRestore(@NonNull Serializable state) {
        ChainSerializable chainSerializable = (ChainSerializable) state;
        super.onRestore(chainSerializable.getState());
        chainSerializable = chainSerializable.getNext();
        fetchInProgress = (boolean) chainSerializable.getState();
        chainSerializable = chainSerializable.getNext();
        lastActionConfirm = (Confirm) chainSerializable.getState();
        chainSerializable = chainSerializable.getNext();
        lastActionPresetId = (int) chainSerializable.getState();
    }

    @SuppressWarnings("ConstantConditions")
    @Nullable
    @Override
    public Serializable onRetain() {
        ChainSerializable chainSerializable = new ChainSerializable(super.onRetain());
        chainSerializable.createNext(fetchInProgress).createNext(lastActionConfirm).createNext(lastActionPresetId);
        return chainSerializable;
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/presenter/impl/HomePresenterImpl.kt
package com.stedi.randomimagegenerator.app.presenter.impl

import android.os.Parcelable
import com.squareup.otto.Subscribe
import com.stedi.randomimagegenerator.app.di.DefaultScheduler
import com.stedi.randomimagegenerator.app.di.UiScheduler
import com.stedi.randomimagegenerator.app.model.data.PendingPreset
import com.stedi.randomimagegenerator.app.model.data.Preset
import com.stedi.randomimagegenerator.app.model.repository.PresetRepository
import com.stedi.randomimagegenerator.app.other.LockedBus
import com.stedi.randomimagegenerator.app.presenter.interfaces.HomePresenter
import rx.Completable
import rx.Scheduler
import rx.Single
import timber.log.Timber
import java.io.Serializable
import javax.inject.Inject

class HomePresenterImpl @Inject constructor(
        private val presetRepository: PresetRepository,
        private val pendingPreset: PendingPreset,
        @DefaultScheduler private val subscribeOn: Scheduler,
        @UiScheduler private val observeOn: Scheduler,
        private val bus: LockedBus) : HomePresenter {

    private var ui: HomePresenter.UIImpl? = null
    private var fetchInProgress: Boolean = false
    private var deleteInProgress: Boolean = false

    private var deletePreset: Preset? = null
    private var generatePreset: Preset? = null

    class FetchPresetsEvent(val presets: List<Preset> = emptyList(), val throwable: Throwable? = null)
    class DeletePresetEvent(val preset: Preset, val throwable: Throwable? = null)

    init {
        bus.register(this)
    }

    override fun onAttach(ui: HomePresenter.UIImpl) {
        this.ui = ui
    }

    override fun onDetach() {
        ui = null
    }

    override fun onDestroy() {
        bus.unregister(this)
    }

    override fun fetchPresets() {
        if (fetchInProgress) {
            return
        }
        fetchInProgress = true

        Timber.d("fetching presets")
        Single.fromCallable { presetRepository.getAll() }
                .subscribeOn(subscribeOn)
                .observeOn(observeOn)
                .subscribe({
                    bus.post(FetchPresetsEvent(it))
                }, { t ->
                    bus.post(FetchPresetsEvent(throwable = t))
                })
    }

    override fun editPreset(preset: Preset) {
        pendingPreset.prepareCandidateFrom(preset)
        ui?.showEditPreset()
    }

    override fun newPreset() {
        pendingPreset.newDefaultCandidate()
        ui?.showCreatePreset()
    }

    override fun deletePreset(preset: Preset) {
        if (deletePreset != null) {
            Timber.d("ignoring deletePreset, because last deletePreset is not finished")
            return
        }

        Timber.d("deletePreset $preset")
        if (pendingPreset.getPreset() === preset) {
            pendingPreset.clearPreset()
            ui?.onPresetDeleted(preset)
            return
        }

        deletePreset = preset
        ui?.showConfirmDeletePreset(preset)
    }

    override fun confirmDeletePreset(confirm: Boolean) {
        val deletePreset = deletePreset
        if (deletePreset == null) {
            return
        }

        if (!confirm) {
            Timber.d("cancelDeletePreset")
            this.deletePreset = null
            return
        }

        Timber.d("confirmDeletePreset")
        val presetId = deletePreset.id
        this.deletePreset = null
        deleteInProgress = true

        Completable.fromCallable {
            presetRepository.remove(presetId)
        }.subscribeOn(subscribeOn)
                .observeOn(observeOn)
                .subscribe({
                    bus.post(DeletePresetEvent(deletePreset))
                }, { t ->
                    bus.post(DeletePresetEvent(deletePreset, t))
                })
    }

    override fun startGeneration(preset: Preset) {
        if (generatePreset != null) {
            Timber.d("ignoring startGeneration, because last startGeneration is not confirmed/canceled")
            return
        }

        Timber.d("startGeneration $preset")
        generatePreset = preset
        ui?.showConfirmGeneratePreset(preset)
    }

    override fun confirmStartGeneration(confirm: Boolean) {
        val generatePreset = generatePreset
        if (generatePreset == null) {
            return
        }

        if (!confirm) {
            Timber.d("cancelStartGeneration")
            this.generatePreset = null
            return
        }

        Timber.d("confirmStartGeneration")
        this.generatePreset = null
        ui?.showGenerationDialog(generatePreset)
    }

    @Subscribe
    fun onFetchPresetsEvent(event: FetchPresetsEvent) {
        if (!fetchInProgress) {
            Timber.d("ignoring FetchPresetsEvent from previous presenter")
            return
        }

        Timber.d("onFetchPresetsEvent")
        fetchInProgress = false

        if (ui == null) {
            Timber.d("onFetchPresetsEvent when ui == null")
            return
        }

        event.throwable?.apply {
            Timber.e(this)
            ui?.onFailedToFetchPresets()
        } ?: ui?.onPresetsFetched(pendingPreset.getPreset(), event.presets)
    }

    @Subscribe
    fun onDeletePresetEvent(event: DeletePresetEvent) {
        if (!deleteInProgress) {
            Timber.d("ignoring DeletePresetEvent from previous presenter")
            return
        }

        Timber.d("onDeletePresetEvent")
        deleteInProgress = false

        if (ui == null) {
            Timber.d("onDeletePresetEvent when ui == null")
            return
        }

        event.throwable?.apply {
            Timber.e(this)
            ui?.onFailedToDeletePreset()
        } ?: ui?.onPresetDeleted(event.preset)
    }

    @Suppress("UNCHECKED_CAST")
    override fun onRestore(state: Serializable) {
        (state as Array<Parcelable?>).apply {
            deletePreset = this[0] as Preset?
            generatePreset = this[1] as Preset?
        }
    }

    override fun onRetain(): Serializable? {
        return arrayOf(deletePreset, generatePreset)
    }
}